### ISSUE TITLE
Batch job fails when using specific schema in PostgreSQL

### ISSUE THREAD

**Issue by grey-scaven** [spring-projects/spring-batch#4104]
**Bug description**
A problem occurs when defining and using meta tables in a specific schema of PostgreSQL.
I used SQL for metatables included in spring batch core. (ex. org/springframework/batch/core/schema-postgresql.sql)

**Environment**
Spring Batch 4.3.5
Java 11
PostgreSQL 12 and metatables on specific schema

**Steps to reproduce**
1. Create spring batch metatables on specific schema (ex. item)
```
CREATE TABLE item.BATCH_JOB_INSTANCE (
	JOB_INSTANCE_ID BIGINT NOT NULL PRIMARY KEY DEFAULT nextval('item.BATCH_JOB_SEQ'),
	VERSION BIGINT,
	JOB_NAME VARCHAR NOT NULL,
	JOB_KEY VARCHAR NOT NULL,
	constraint JOB_INST_UN unique (JOB_NAME, JOB_KEY)
);

```
2. Running batch until 30 times
3. An error occurs and aborted as below
```
Caused by: org.postgresql.util.PSQLException: ERROR: insert or update on table "batch_job_execution" violates foreign key constraint "job_inst_exec_fk"
  Detail: Key (job_instance_id)=(30) is not present in table "batch_job_instance".
```

**Expected behavior**
Batch job doesn't fail because of invalid foreign key constraint.

**Comment by fmbenhassine** [comment]
Thank you for opening this issue. Running the [PostgreSQLJobRepositoryIntegrationTests](https://github.com/spring-projects/spring-batch/blob/main/spring-batch-core/src/test/java/org/springframework/batch/core/test/repository/PostgreSQLJobRepositoryIntegrationTests.java) test repeatedly 30 times does not fail with the reported exception. I tried with both Postgres v12.13 and v13.3.

So the issue you are reporting does not seem to be related to the way the foreign key constraint is declared as mentioned here and changed in #4105.

Do you have a concurrent execution that might cause the issue? Please provide a failing test that reproduces the issue and we will dig deeper. Thank you upfront.

**Comment by SiegfredRodriguez** [comment]
I am also encountering this issue on 5.0.0.

**Comment by fmbenhassine** [comment]
@SiegfredRodriguez Thank for letting us know that this issue happens with 5.0.0 as well. I don't think the Spring Batch version plays a role here.

Do you have a concurrent execution that might cause the issue? 

I shared a full integration test in my previous comment, so if you manage to update it according to your use case and show how it fails, that would help us understand the root cause of the issue. Thank you for your collaboration.

**Comment by rkrepos** [comment]
I am facing a similar issue. I have a very simple spring boot application. JDBC Postgres datasource in application.yml. Batch configuration using @EnableBatchProcessing and uses the default JobBuilderFactory, StepBuilderFactory and JobLauncher beans. Just a single job with one tasklet step to read db and write to a file. No extra datasource or custom configurations for spring batch.

Batch metadata tables were  automatically created in the custom schema used by the application. Ran fine for 30 times and then started throwing below exception.

PreparedStatementCallback; bad SQL grammar [SELECT JOB_INSTANCE_ID, JOB_NAME from BATCH_JOB_INSTANCE where JOB_NAME = ? and JOB_KEY = ?]; nested exception is org.postgresql.util.PSQLException: ERROR: relation "batch_job_instance" does not exist.

Job was launched by a cron expression to run every minute. Except the first time, there was no new data in the DB. So the job just ran for few milliseconds each time. Also, there is no DB read/write outside the job. 

To rule out connection timeout, I changed the cron expression to every 10 minutes, and the exception occurred on the 4th launch of the job, thus confirming it is  a 30 minutes issue and not 30 counts. Customized the Hikari pool configuration in the yaml to "hikari:max-lifetime: 300000". Now the exception throws after 5 minutes. Even added a "keepalive-time: 150000", but it didn't help.

**Comment by rkrepos** [comment]
By the way, even though I didn't think it makes a difference,  I tried the constraint change as per https://github.com/spring-projects/spring-batch/pull/4105. As expected it didn't fix the issue. The exception is thrown after the datasource pool connection maxLifetime.

**Comment by fmbenhassine** [comment]
@rkrepos Thank you for your feedback, I came to the same conclusion as well when I tested that PR on my side. Could you please add a comment on that PR to say that you tried it but it did not fix the issue? Thank you upfront.

**Comment by rkrepos** [comment]
@fmbenhassine, would you please be able to advise on how to best to handle this? As mentioned earlier, I am using the default JobBuilderFactory, StepBuilderFactory and JobLauncher beans. Currently I am running the application with "hikari:max-lifetime: 0" which is not desirable.

**Comment by fmbenhassine** [comment]
I would love to help, but without a minimal example that reproduces the issue I cannot understand the problem and advise how to handle it.

I shared a full integration test in my previous comment, so if you can update it to make it fail or provide a minimal example that reproduces the issue I will take a look. See [Issue Reporting Guidelines](https://github.com/spring-projects/spring-batch/blob/main/ISSUE_REPORTING.md) for a starting point.

**Comment by grey-scaven** [comment]
If you need to create a batch meta table in a specific schema and propose changes to foreign key constraints, the DDL for the order schema would need to be structured as follows. I have specified only the parts that are not dynamically changing

```
CREATE TABLE order.BATCH_JOB_EXECUTION  (
	JOB_EXECUTION_ID BIGINT  NOT NULL PRIMARY KEY ,
	VERSION BIGINT  ,
	JOB_INSTANCE_ID BIGINT NOT NULL,
	JOB_INSTANCE_ID BIGINT NOT NULL
		constraint JOB_INST_EXEC_FK
			references BATCH_JOB_INSTANCE,
	CREATE_TIME TIMESTAMP NOT NULL,
	START_TIME TIMESTAMP DEFAULT NULL ,
	END_TIME TIMESTAMP DEFAULT NULL ,
	STATUS VARCHAR(10) ,
	EXIT_CODE VARCHAR(2500) ,
	EXIT_MESSAGE VARCHAR(2500) ,
	LAST_UPDATED TIMESTAMP,
	constraint JOB_INST_EXEC_FK foreign key (JOB_INSTANCE_ID)
	references BATCH_JOB_INSTANCE(JOB_INSTANCE_ID)
	LAST_UPDATED TIMESTAMP
) ;
```

**Comment by rkrepos** [comment]
@grey-scaven, I had tried the #4105 fix across all the batch metadata tables and it didn't solve the issue. Secondly, the syntax above is incorrect I am afraid. JOB_INSTANCE_ID is replicated. 

Also my observation was that it is not a count related issue but rather a 30 minutes connection timeout issue with the datasource connection pool, which was default for the hikari pool that my spring boot application uses.

Finally, at least in my case, it didn't seem to be due to a bug in spring batch. I suspected that spring batch, after the connection timeout, and with a new connection from the pool, was attempting to access public schema instead of the custom schema. To test this, I created the batch tables in public schema as well (in addition to the custom ones) but with a different owner. This time, after the timeout, spring batch started complaining about permission issues instead of missing foreign key.

In conclusion, the issue was with hikari connection pool and not spring batch. Further reading pointed towards https://github.com/brettwooldridge/HikariCP/issues/1633. As per this report, hikari pool switches to public schema after the connection closure for Postgres. One workaround is to add "?currentSchema=CustomSchemaName" with the jdbc connection url. This has solved the issue with my application.

**Comment by grey-scaven** [comment]
@rkrepos Thank you for reporting.
